File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/ClosestPair.java
package com.divideconquer.algorithms;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

public final class ClosestPair {
    private ClosestPair() {
    }

    public static final class Point {
        private final double x;
        private final double y;

        public Point(double x, double y) {
            this.x = x;
            this.y = y;
        }

        public double x() {
            return x;
        }

        public double y() {
            return y;
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) {
                return true;
            }
            if (!(other instanceof Point)) {
                return false;
            }
            Point point = (Point) other;
            return Double.compare(point.x, x) == 0 && Double.compare(point.y, y) == 0;
        }

        @Override
        public int hashCode() {
            return Objects.hash(x, y);
        }

        @Override
        public String toString() {
            return "(" + x + ", " + y + ")";
        }
    }

    public static final class Result {
        private final Point first;
        private final Point second;
        private final double distance;

        public Result(Point first, Point second, double distance) {
            this.first = first;
            this.second = second;
            this.distance = distance;
        }

        public Point first() {
            return first;
        }

        public Point second() {
            return second;
        }

        public double distance() {
            return distance;
        }
    }

    public static Result find(List<Point> points) {
        if (points == null || points.size() < 2) {
            throw new IllegalArgumentException("at least two points are required");
        }
        if (points.stream().anyMatch(point -> point == null)) {
            throw new IllegalArgumentException("points must not contain null values");
        }
        List<Point> byX = new ArrayList<>(points);
        byX.sort(Comparator.comparingDouble(Point::x).thenComparingDouble(Point::y));
        List<Point> byY = new ArrayList<>(points);
        byY.sort(Comparator.comparingDouble(Point::y).thenComparingDouble(Point::x));
        return closest(byX, byY);
    }

    private static Result closest(List<Point> byX, List<Point> byY) {
        int n = byX.size();
        if (n <= 3) {
            return bruteForce(byX);
        }

        int mid = n / 2;
        Point midPoint = byX.get(mid);
        List<Point> leftX = new ArrayList<>(byX.subList(0, mid));
        List<Point> rightX = new ArrayList<>(byX.subList(mid, n));
        List<Point> leftY = new ArrayList<>();
        List<Point> rightY = new ArrayList<>();

        for (Point point : byY) {
            if (point.x() < midPoint.x() || (point.x() == midPoint.x() && leftY.size() < leftX.size())) {
                leftY.add(point);
            } else {
                rightY.add(point);
            }
        }

        Result left = closest(leftX, leftY);
        Result right = closest(rightX, rightY);
        Result best = left.distance() <= right.distance() ? left : right;

        List<Point> strip = new ArrayList<>();
        for (Point point : byY) {
            if (Math.abs(point.x() - midPoint.x()) < best.distance()) {
                strip.add(point);
            }
        }

        for (int i = 0; i < strip.size(); i++) {
            for (int j = i + 1; j < strip.size() && strip.get(j).y() - strip.get(i).y() < best.distance(); j++) {
                Result candidate = result(strip.get(i), strip.get(j));
                if (candidate.distance() < best.distance()) {
                    best = candidate;
                }
            }
        }
        return best;
    }

    private static Result bruteForce(List<Point> points) {
        Result best = result(points.get(0), points.get(1));
        for (int i = 0; i < points.size(); i++) {
            for (int j = i + 1; j < points.size(); j++) {
                Result candidate = result(points.get(i), points.get(j));
                if (candidate.distance() < best.distance()) {
                    best = candidate;
                }
            }
        }
        return best;
    }

    private static Result result(Point a, Point b) {
        double dx = a.x() - b.x();
        double dy = a.y() - b.y();
        return new Result(a, b, Math.hypot(dx, dy));
    }
}

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/closest_pair.py
"""Closest pair of points in O(n log n)."""

from __future__ import annotations

from dataclasses import dataclass
from math import hypot


@dataclass(frozen=True, order=True)
class Point:
    x: float
    y: float


@dataclass(frozen=True)
class Result:
    first: Point
    second: Point
    distance: float


def find(points: list[Point]) -> Result:
    """Return the closest pair among at least two points."""
    if points is None:
        raise ValueError("points must not be None")
    if len(points) < 2:
        raise ValueError("at least two points are required")
    if any(point is None for point in points):
        raise ValueError("points must not contain None values")
    by_x = sorted(points, key=lambda point: (point.x, point.y))
    by_y = sorted(points, key=lambda point: (point.y, point.x))
    return _closest(by_x, by_y)


def _closest(by_x: list[Point], by_y: list[Point]) -> Result:
    if len(by_x) <= 3:
        return _brute_force(by_x)

    mid = len(by_x) // 2
    mid_point = by_x[mid]
    left_x = by_x[:mid]
    right_x = by_x[mid:]
    left_lookup = set(left_x)
    left_y = [point for point in by_y if point in left_lookup]
    right_y = [point for point in by_y if point not in left_lookup]

    left = _closest(left_x, left_y)
    right = _closest(right_x, right_y)
    best = left if left.distance <= right.distance else right

    strip = [point for point in by_y if abs(point.x - mid_point.x) < best.distance]
    for i, point in enumerate(strip):
        for other in strip[i + 1 : i + 8]:
            if other.y - point.y >= best.distance:
                break
            candidate = _result(point, other)
            if candidate.distance < best.distance:
                best = candidate
    return best


def _brute_force(points: list[Point]) -> Result:
    best = _result(points[0], points[1])
    for i, point in enumerate(points):
        for other in points[i + 1 :]:
            candidate = _result(point, other)
            if candidate.distance < best.distance:
                best = candidate
    return best


def _result(a: Point, b: Point) -> Result:
    return Result(a, b, hypot(a.x - b.x, a.y - b.y))
